src/components/alerts/BannerCarousel/BannerCarousel.tsx

```diff
@@ -1,39 +1,267 @@
-import * as React from 'react';
+import classnames from 'classnames';
+import React, { Children, useEffect, useLayoutEffect, useRef, useState } from 'react';
 import IReactComponentProps from '../../../common/structures/IReactComponentProps';
+import styles from './BannerCarousel.scss';
 
-interface IState {
-	currentIndex?: number;
-}
+type Direction = 'up' | 'down';
+type ChildrenArray = ReturnType<typeof Children['toArray']>;
 
-export default class BannerCarousel extends React.Component<IReactComponentProps, IState> {
-	constructor (props: IReactComponentProps) {
-		super(props);
+const BannerCarousel = (props: IReactComponentProps) => {
+	const { children, className, id, style } = props;
 
-		this.state = {
-			currentIndex: 0,
-		};
-	}
+	// Dummy banner ref used to set height
+	const bannerRef = useRef<HTMLDivElement>(null);
 
-	render () {
-		return (
+	// State
+	const [childrenArray, setChildrenArray] = useState(Children.toArray(children));
+	const [numBanners, setNumBanners] = useState(childrenArray.length);
+	const [currentIndex, setCurrentIndex] = useState(childrenArray.length ? 0 : -1);
+	const [currentBanner, setCurrentBanner] = useState<React.ReactNode>(null);
+	const [nextBanner, setNextBanner] = useState<React.ReactNode>(null);
+	const [switching, setSwitching] = useState(false);
+	const [switchDirection, setSwitchDirection] = useState<Direction>('up');
+	const [carouselHeight, setCarouselHeight] = useState(0);
+	const [gettingHeight, setGettingHeight] = useState(false);
+	const [activeTimeout, setActiveTimeout] = useState<NodeJS.Timeout>(0 as unknown as NodeJS.Timeout);
+
+	// Need to safely set these banners in state with unique keys to prevent React errors
+	const safeSetNextBanner = (banner: React.ReactNode) => {
+		if (React.isValidElement(banner)) {
+			setNextBanner(React.cloneElement(banner, { key: 'next-banner' }));
+		} else {
+			setNextBanner(null);
+		}
+	};
+
+	const safeSetCurrentBanner = (banner: React.ReactNode) => {
+		if (React.isValidElement(banner)) {
+			setCurrentBanner(React.cloneElement(banner, { key: 'current-banner' }));
+		} else {
+			setCurrentBanner(null);
+		}
+	};
+
+	// Set up inital banners and height
+	useEffect(() => {
+		safeSetCurrentBanner(childrenArray[0]);
+		safeSetNextBanner(childrenArray[0]);
+		setGettingHeight(true);
+
+		return () => clearTimeout(activeTimeout);
+	}, []);
+
+	// useLayoutEffect to grab height of dummy banner BEFORE the browser paints
+	// allows us to set the height without visual glitches
+	useLayoutEffect(() => {
+		if (gettingHeight) {
+			// Carousel button height will equal 13px per unselected banner, plus 20px for selected, plus 20px padding
+			const buttonsHeight = numBanners > 1 ? (numBanners - 1) * 13 + 20 + 20 : 0;
```

Comment:
Yeah this was to avoid using another `ref` for the carousel. Could separate out the numbers into named variables to make it clearer, but I'm inclined to keep the math at least since it works pretty well